dataset_loaders/videos/movingMNIST.py

```diff
@@ -0,0 +1,222 @@
+                 image_size=64, num_digits=1, digits_sizes=[28, 28],
+                 digits_speed=0.3, change_dir_prob=[0., 0.],
+                 rng=None, seed=1, *args, **kwargs):
+
+        self.data_shape = (image_size,) * 2 + (self.data_shape[2],)
+
+        self.which_set = 'validation' if which_set == 'valid' else which_set
+        self.nvids = nvids
+        self.image_size_ = image_size
+        self.row_ = 0
+        self.num_digits_ = num_digits
+        self.num_channels_ = 1
+        self.digits_sizes_ = digits_sizes
+        self.num_digits_ = num_digits
+        self.batch_size_ = 1
+        self.set_has_GT = False
+        self.seed = seed
+        self.digits_speed = digits_speed
+        self.change_dir_prob = change_dir_prob
+        self.indices_ = range(nvids)
+        self._rng = rng if rng else np.random.RandomState(seed)
+        # self.path = self.shared_path
+        import h5py
+        try:
+            f = h5py.File(os.path.join(self.path, 'mnist.h5'))
+        except:
+            raise RuntimeError('Failed to load dataset file')
+
+        self.data_ = f[self.which_set].value[:nvids].reshape(-1, 28, 28)
+        f.close()
+
+        super(MovingMNISTDataset, self).__init__(*args, **kwargs)
+
+    def get_names(self):
+        """Return a dict of names, per prefix/subset."""
+        return {'default': range(self.nvids*self.seq_length)}
+
+    def Overlap(self, a, b):
+        """Put b on top of a."""
+        return np.maximum(a, b)
+
+    def get_random_trajectory(self, batch_size):
+        length = self.seq_length + 1
+        canvas_size = self.image_size_ - np.max(self.digits_sizes_)
+
+        # Initial position uniform random inside the box.
+        y = self._rng.rand(batch_size)
+        x = self._rng.rand(batch_size)
+        # Choose speed.
+        speed = self.digits_speed
+        if isinstance(self.digits_speed, float):
+            speed = [self.digits_speed for el in range(batch_size)]
+        elif self.digits_speed == 'random':
+            speed = self._rng.rand(batch_size)
+        v_y = np.asarray(speed)  # np.sin(theta)
+        v_x = np.asarray(speed)  # np.cos(theta)
+
+        start_y = np.zeros((length, batch_size))
+        start_x = np.zeros((length, batch_size))
+        # Random directions
+        y_rand = np.asarray([self._rng.uniform(-1, 1) for _ in
+                             range(batch_size)])
+        x_rand = np.asarray([self._rng.uniform(-1, 1) for _ in
+                             range(batch_size)])
+
+        # For each digit for each frame compute the decision to change
+        # direction following a binomial distribution
+        change_dir = []
+        for i in range(self.num_digits_):
+            sampling = [self._rng.binomial(
+                1, self.change_dir_prob[i]) for _ in range(length)]
+            change_dir.append(sampling)
+        change_dir = zip(*change_dir)
+
+        for i in xrange(length):
```

Comment:
Why `xrange`? It's not such a big list and it's not python3 compatible (should we switch). Replace with `range`